tls: prefer path over port in connect

bengl opened the pull request:
Makes tls.connect() behave as documented, preferring options.path over
options.port. This makes it consistent with net.connect(), so the
included test demonstrates that both behave in this way.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
tls, test

bengl commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9429/

mscdex commented:
Is this semver-major?

bengl commented:
Yes most likely semver major.

jasnell commented:
ping @nodejs/ctc

BridgeAR commented:
If it is documented that path is preferred over port and it is also consistent towards net.connect(), I would argue it is a semver-patch. Especially as the options are not compatible with each other and should normally not coexist in the first place.

thefourtheye commented on test/parallel/test-tls-net-connect-prefer-path.js:
```diff
@@ -0,0 +1,66 @@
+        path: unixServer.address(),
+        port: tcpServer.address().port,
+        host: 'localhost',
+        rejectUnauthorized: false,
```
Nit: extra comma at the end.

thefourtheye commented on test/parallel/test-tls-net-connect-prefer-path.js:
```diff
@@ -0,0 +1,66 @@
+  });
+}
+
+testLib(net, () => testLib(tls, common.mustCall()));
```
Callback for `net` case also should have `common.mustCall`

thefourtheye commented on test/parallel/test-tls-net-connect-prefer-path.js:
```diff
@@ -0,0 +1,66 @@
+    });
+  }
+  const server = lib.createServer(...args);
+  server.listen(tcp ? 0 : common.PIPE, () => cb(server));
```
`common.mustCall` for this callback also?

thefourtheye commented on test/parallel/test-tls-net-connect-prefer-path.js:
```diff
@@ -0,0 +1,66 @@
+        rejectUnauthorized: false,
+      }, () => {
+        const bufs = [];
+        client.on('data', common.mustCall((d) => {
```
Generally, `data` handler will be called multiple times. If we are sure that it will be called only once, then `bufs` need not be an array, right?

bengl replied:
I don't think there's a reason to be *sure* it will be only called once, is there?

thefourtheye commented on test/parallel/test-tls-net-connect-prefer-path.js:
```diff
@@ -0,0 +1,66 @@
+          assert.strictEqual(`${libName(lib)}:${unixServer.address()}`, resp);
+          tcpServer.close();
+          unixServer.close();
+          if (cb) {
```
All the tests pass a callback function. Why do we need this check?

thefourtheye commented on lib/_tls_wrap.js:
```diff
@@ -1056,18 +1056,14 @@ exports.connect = function(...args /* [port,] [host,] [options,] [cb] */) {
-        lookup: options.lookup
-      };
-    }
+    const connect_opt = {
```
Nit: I know that this uses the same variable name as the old one, but most part of the core uses camel case. Could you change this?

thefourtheye commented on lib/_tls_wrap.js:
```diff
@@ -1056,18 +1056,14 @@ exports.connect = function(...args /* [port,] [host,] [options,] [cb] */) {
-      };
-    }
+    const connect_opt = {
+      path: options.path,
```
I would prefer a documentation change, which explicitly states that if `path` is present, that gets first preference.

bengl replied:
That's already stated in the `tls` doc. I'll add it to `net`.

BridgeAR commented:
Ping @bengl

bengl commented:
Whoops! Got swamped with product work and let this slide. I'll try and clean this up in the next few days.

bengl commented:
@thefourtheye latest commit addresses most of your nits I think, PTAL. Also note the doc change, as requested.

New CI: https://ci.nodejs.org/job/node-test-pull-request/10152/

bengl commented:
On further thought, I agree with @BridgeAR here, in that this is semver-patch, since it's making the behavior match the existing documentation, while it didn't before. I'll switch the label.

BridgeAR commented:
@nodejs/tsc is everyone fine with this being a semver-patch?

jasnell commented:
If there are no objections from tsc by friday it should be ok to proceed with semver-patch

bengl commented:
OK, cool. It's Friday. I'll land this shortly.

bengl commented:
Landed in 6f1caadb85f70bb902c078d9b7a96e3c4b36b31f.

mcollina commented:
This broke MQTT.js https://github.com/mqttjs/MQTT.js/issues/699 and https://github.com/mqttjs/MQTT.js/pull/700. It was indeed semver-major.
I do not think we should revert, as the original behavior was undocumented.

MylesBorins commented:
I've opted to not land this on 6.x due to the behavior change.

@mcollina lmk if you think this should be backported
